controllers/rollingupgrade_controller.go

```diff
@@ -704,9 +704,50 @@ func (r *RollingUpgradeReconciler) Process(ctx *context.Context,
 		return
 	}
 
+	//Validation step: check if all the nodes have the latest launchconfig.
+	r.info(ruObj, "Validating the launch definition of nodes and ASG")
+	if err := r.validateNodesLaunchDefinition(ruObj); err != nil {
+		r.error(ruObj, err, "Launch definition validation failed")
+		r.finishExecution(StatusError, nodesProcessed, ctx, ruObj)
+		return
+	}
+	r.info(ruObj, "Launch definition validation succeeded")
+
 	r.finishExecution(StatusComplete, nodesProcessed, ctx, ruObj)
 }
 
+//Check if ec2Instances and the ASG have same launch config.
+func (r *RollingUpgradeReconciler) validateNodesLaunchDefinition(ruObj *upgrademgrv1alpha1.RollingUpgrade) error {
+	//Get ASG launch config
+	var err error
+	err = r.populateAsg(ruObj)
+	if err != nil {
+		return errors.New("Unable to populate the ASG object")
+	}
+	asg, err := r.GetAutoScalingGroup(ruObj.Name)
+	if err != nil {
+		return fmt.Errorf("Unable to load ASG with name: %s", ruObj.Name)
+	}
+	launchDefinition := NewLaunchDefinition(asg)
+	launchConfigASG, launchTemplateASG := launchDefinition.launchConfigurationName, launchDefinition.launchTemplate
+
+	//Get ec2 instances and their launch configs.
+	ec2instances := asg.Instances
+	for _, ec2Instance := range ec2instances {
+		ec2InstanceID, ec2InstanceLaunchConfig, ec2InstanceLaunchTemplate := ec2Instance.InstanceId, ec2Instance.LaunchConfigurationName, ec2Instance.LaunchTemplate
+		if aws.StringValue(launchConfigASG) != aws.StringValue(ec2InstanceLaunchConfig) {
+			r.info(ruObj, "Launch config doesn't match with ASG", "Instance ID", aws.StringValue(ec2InstanceID), "Instance LaunchConfig", aws.StringValue(ec2InstanceLaunchConfig))
+			err = errors.New("Launch config mismatch")
```

Comment:
There is no need to log an error (line 739) if you are returning an error (it will be logged downstream).
Also, why not just return the error at this point?
In addition, can we use a more descriptive message here?
e.g. "post-validation failed: instances found with mismatching launch configuration" 

Same for launch template below

Reply:
I thought it might be useful to know which all instances have a mismatch. Hence returning at the end. True. The logging here was also for the same reason(I included the instance ID int the logs)

Reply:
If you're relying on upstream logging, the error would be logged as message so you could be detailed like so...
```suggestion
			err = fmt.Errorf("launch config mismatch, instance config %s does not match asg config %s", aws.StringValue(ec2InstanceID), aws.StringValue(ec2InstanceLaunchConfig))
```
Also, go errors are not supposed to have capitalization. See https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Another issue is your in a loop and return here is outside the loop, so it's correct that `return err` should be done within the if blocks or you should collect all errors and return a concatenated list or the complete error slice `[]err`

Reply:
Ah! I see. I shall incorporate these. Thanks, Kevin!